flight/PiOS/Common/pios_port.c

```diff
@@ -0,0 +1,623 @@
+					PIOS_Assert(0);
+				}
+			}
+#endif  /* PIOS_INCLUDE_COM */
+			break;
+	}
+}
+#endif
+
+#if defined(PIOS_INCLUDE_USB_HID)
+void PIOS_HAL_configure_HID(HwSharedUSB_HIDPortOptions port_type,
+		uintptr_t usb_id,
+		const struct pios_usb_hid_cfg *hid_cfg) {
+	uintptr_t pios_usb_hid_id;
+	if (PIOS_USB_HID_Init(&pios_usb_hid_id, hid_cfg, usb_id)) {
+		PIOS_Assert(0);
+	}
+
+	switch (port_type) {
+		case HWSHARED_USB_HIDPORT_DISABLED:
+			break;
+		case HWSHARED_USB_HIDPORT_USBTELEMETRY:
+#if defined(PIOS_INCLUDE_COM)
+			{
+				uint8_t * rx_buffer = (uint8_t *) PIOS_malloc(PIOS_COM_TELEM_USB_RX_BUF_LEN);
+				uint8_t * tx_buffer = (uint8_t *) PIOS_malloc(PIOS_COM_TELEM_USB_TX_BUF_LEN);
+				PIOS_Assert(rx_buffer);
+				PIOS_Assert(tx_buffer);
+				if (PIOS_COM_Init(&pios_com_telem_usb_id, &pios_usb_hid_com_driver, pios_usb_hid_id,
+							rx_buffer, PIOS_COM_TELEM_USB_RX_BUF_LEN,
+							tx_buffer, PIOS_COM_TELEM_USB_TX_BUF_LEN)) {
+					PIOS_Assert(0);
+				}
+			}
+#endif  /* PIOS_INCLUDE_COM */
+			break;
+		case HWSHARED_USB_HIDPORT_RCTRANSMITTER:
+#if defined(PIOS_INCLUDE_USB_RCTX)
+			{
+				if (PIOS_USB_RCTX_Init(&pios_usb_rctx_id, &pios_usb_rctx_cfg, pios_usb_id)) {
+					PIOS_Assert(0);
+				}
+			}
+#endif  /* PIOS_INCLUDE_USB_RCTX */
+			break;
+
+	}
+}
+#endif  /* PIOS_INCLUDE_USB_HID */
+
+
+#if defined(PIOS_INCLUDE_RFM22B)
+void PIOS_HAL_configure_RFM22B(HwSharedRadioPortOptions radio_type,
+		uint8_t board_type, uint8_t board_rev,
+		HwSharedMaxRfPowerOptions max_power,
+		HwSharedMaxRfSpeedOptions max_speed,
+		const struct pios_openlrs_cfg *openlrs_cfg,
+		const struct pios_rfm22b_cfg *rfm22b_cfg,
+		uint8_t min_chan, uint8_t max_chan, uint32_t coord_id) {
+	/* Initalize the RFM22B radio COM device. */
+	RFM22BStatusInitialize();
+	RFM22BStatusCreateInstance();
+
+	RFM22BStatusData rfm22bstatus;
+	RFM22BStatusGet(&rfm22bstatus);
+	RFM22BStatusInstSet(1,&rfm22bstatus);
+
+	// Initialize out status object.
+	rfm22bstatus.BoardType     = board_type;
+	rfm22bstatus.BoardRevision = board_rev;
+
+	if (radio_type == HWSHARED_RADIOPORT_OPENLRS) {
+#if defined(PIOS_INCLUDE_OPENLRS_RCVR)
+		uintptr_t openlrs_id;
+
+		PIOS_OpenLRS_Init(&openlrs_id, PIOS_RFM22_SPI_PORT, 0, openlrs_cfg);
+
+		{
+			uintptr_t pios_rfm22brcvr_id;
+			PIOS_OpenLRS_Rcvr_Init(&pios_rfm22brcvr_id, openlrs_id);
+			uintptr_t pios_rfm22brcvr_rcvr_id;
+			if (PIOS_RCVR_Init(&pios_rfm22brcvr_rcvr_id, &pios_openlrs_rcvr_driver, pios_rfm22brcvr_id)) {
+				PIOS_Assert(0);
+			}
+			pios_rcvr_group_map[MANUALCONTROLSETTINGS_CHANNELGROUPS_OPENLRS] = pios_rfm22brcvr_rcvr_id;
+		}
+#endif /* PIOS_INCLUDE_OPENLRS_RCVR */
+	} else if (radio_type == HWSHARED_RADIOPORT_DISABLED ||
+			max_power == HWSHARED_MAXRFPOWER_0) {
+		// When radio disabled, it is ok for init to fail. This allows
+		// boards without populating this component.
+		if (PIOS_RFM22B_Init(&pios_rfm22b_id, PIOS_RFM22_SPI_PORT, rfm22b_cfg->slave_num, rfm22b_cfg) == 0) {
```

Comment:
this is the bit that should be reworked for not reininting the modem when not in use. I need a different method that PIOS_RFM22B_Init. I won't fix it now since it will clash with your work.